src/util/Timer.cpp

```diff
@@ -239,6 +240,104 @@ VirtualClock::setCurrentVirtualTime(time_point t)
     mVirtualNow = t;
 }
 
+void
+VirtualClock::sleep_for(std::chrono::microseconds us)
+{
+    if (mMode == VIRTUAL_TIME)
+    {
+        setCurrentVirtualTime(now() + us);
+    }
+    else
+    {
+        std::this_thread::sleep_for(us);
+    }
+}
+
+void
+VirtualClock::advanceExecutionQueue()
+{
+    // limit in terms of number of items
+    // ideally we would just pick up everything we can up to the duration.
+    // There is a potential issue with picking up too many items all the time,
+    // in that round robin gets defeated if work is submitted in batches.
+    size_t EXECUTION_QUEUE_BATCH_SIZE = 100;
+
+    // DURATION here needs to be small enough that timers get to fire often
+    // enough but it needs to be big enough so that we do enough work generated
+    // by pulling items from the IO queue
+    std::chrono::milliseconds MAX_EXECUTION_QUEUE_DURATION{500};
+
+    // safety in case we can't keep up at all with work
+    const size_t QUEUE_SIZE_DROP_THRESHOLD = 10000;
+    if (mExecutionQueueSize > QUEUE_SIZE_DROP_THRESHOLD)
+    {
+        // 1. drain elements that can be dropped
+        for (auto it = mExecutionQueue.begin(); it != mExecutionQueue.end();)
+        {
+            if (it->first.mDropable)
+            {
+                mExecutionQueueSize -= it->second.size();
+                it = mExecutionQueue.erase(it);
+            }
+            else
+            {
+                ++it;
+            }
+        }
+        // 2. update parameters so that we help drain the queue
+        // this may lock up the instance for a while, but this is the best thing
+        // we can do
+        EXECUTION_QUEUE_BATCH_SIZE = mExecutionQueueSize;
```

Comment:
Maybe? What I was assuming here is that if the other queue grew so much, we probably have created debt for the rest, so we need to somehow pay it back.

If we do it like that, I think the way it can degenerate is:
* drop a few txs, continue as normal <<< repeat this for a while
* finally, hit a debt so high that we have to process `QUEUE_SIZE_DROP_THRESHOLD` of non droppable items (which are typically expensive to run)